### PR TITLE
Support for # anf &feature, plus an upgrade to packages

### PR DESCRIPTION
**Pull request by mikkopori**
Hello,

here is a suggestion for a couple of improvements. Hope they pass the tests as I was unable to run them, but I did run tests of our which were successful for different formats.

Best regards,

Mikko Pöri

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -115,6 +115,11 @@ function vine(str) {
  * @returns {string|undefined}
  */
 function youtube(str) {
+
+	// remove time and feature
+	str = str.replace(/#t=.*/, "");
+	str = str.replace(/&feature=.*/, "");
```
Comment (radiovisual) [inline index.js]: The last part of this regex looks a bit destructive (or greedy) in the sense that it will try to remove **everything** after the `feature=`, which might be unsafe, since we have any other routines that need to search for url parameters (see some of the functions below this one for examples like `v` or `vi`). So you would want to tighten up this RegExp and add some unit tests to cover all these edge cases.

File: test.js [hunk 2]
```diff
@@ -284,6 +284,24 @@ test('removes -nocookie', t => {
 	t.is(fn('http://www.youtube-nocookie.com/user/username#p/u/1/ABC12302').service, 'youtube');
 });
 
+/** Unfortunately I was not able to run the tests locally as there were
```
Comment (radiovisual) [inline test.js]: I would rather not see a code comment about how you couldn't run the tests baked into the source file itself. We should probably work out the reasons why you couldn't run the tests and fix those. Where were all these syntax errors coming from?

File: test.js [hunk 3]
```diff
@@ -284,6 +284,24 @@ test('removes -nocookie', t => {
...
+	t.is(fn('https://youtube.com/user/WMFinland#p/a/u/1/G-3YxlZIhus').id, 'G-3YxlZIhus');
+
+	t.is(fn('http://www.youtube.com/watch?v=G-3YxlZIhus#t=0m10s').service, 'youtube');
+});
+
+test('removes feature', t => {
+	t.is(fn('http://youtu.be/G-3YxlZIhus&feature=channel').id, 'G-3YxlZIhus');
+	t.is(fn('http://youtube.com/vi/G-3YxlZIhus&feature=channel').id, 'G-3YxlZIhus');
+
+	t.is(fn('http://youtube.com/vi/G-3YxlZIhus&feature=channel').service, 'youtube');
```
Comment (radiovisual) [inline test.js]: Let's add a test that looks like this so that we can make sure that whatever routine you use to remove the `feature` parameter doesn't break the search for the videoid

```js
t.is(fn('http://www.youtube.com/watch?feature=foo&v=G-3YxlZIhus#t=0m10s').id, 'G-3YxlZIhus');
t.is(fn('http://www.youtube.com/watch?feature=foo&v=G-3YxlZIhus#t=0m10s').service, 'youtube');
```

File: test.js [hunk 4]
```diff
@@ -284,6 +284,24 @@ test('removes -nocookie', t => {
 	t.is(fn('http://www.youtube-nocookie.com/user/username#p/u/1/ABC12302').service, 'youtube');
 });
 
+/** Unfortunately I was not able to run the tests locally as there were
+64 different syntax errors, but I've run our own tests with the same urls
+**/
+test('removes #t=0m10s ', t => {
```
Comment (radiovisual) [inline test.js]:
```suggestion
test('removes time hash #t', t => {
```

File: index.js [hunk 5]
```diff
@@ -115,6 +115,11 @@ function vine(str) {
  * @returns {string|undefined}
  */
 function youtube(str) {
+
+	// remove time and feature
+	str = str.replace(/#t=.*/, "");
```
Comment (radiovisual) [inline index.js]: Maybe we should put in some protection here, because the greediness of this regex makes me nervous (where it will remove *everything* after the `#t=`) I think in general we should probably be safe to assume that the time parameter comes at the very end of the url, but for future proofing, and general saftey, maybe for now we should add an `$` at the end of the regex to denote the "end of the string" just to be sure that we are removing the time parameter from the end of the url 

```suggestion
	str = str.replace(/#t=.*$/, "");
```

File: index.js (hunk 1 above)
Comment (radiovisual) [reply index.js:121]: Also, it's not clear to me what value is provided by removing these specific url parameters. Can you elaborate @mikkopori?

Comment (radiovisual) [reply index.js:121]: Nevermind, I see the value now that I ran your tests, without your implementation 

image

Comment (radiovisual) [reply index.js:121]: But this RegEx is too tight. It will fail other url patterns that include `feature` like: 

```
http://www.youtube.com/attribution_link?a=fF1CWYwxCQ4&feature=em-uploademail&u=/watch?v=ABC12305
```

image

Comment (radiovisual) [reply index.js:121]: It might be safer to remove the feature parameter after the id is extracted:

**Before**:

```
G-3YxlZIhus&feature=channel
```

**After**:

```
G-3YxlZIhus
```

**Comment by radiovisual** [comment]
@mikkopori I am closing this PR, since these changes were made and merged in #37. I am sorry I didn't get a chance to merge your contributions. But there were too many open comments that weren't addressed, and right now I have some rare free time to update this package. So the bright side here is that the features you need are now part of this package. So thanks very much for your contributions, but I am sorry that it didn't end up being your PR that was merged. If you feel like there is still something missing from the work I did in #37, please open a new PR, or submit a new feature request. Thanks!

**Comment by mikkopori** [comment]
No, this is fine. And sorry on my part that I never got back to the issues you raised :)

When do you think you will get a new release together?

**Comment by radiovisual** [comment]
@mikkopori This is now live in version `3.1.5`
